### PR TITLE
Lina/settle time visualization

### PR DESCRIPTION
**Pull request by lin99**
Addint settleTime visualization for images

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,18 +18,20 @@
...
   "repository": {
     "type": "git",
-    "url": "https://github.com/Microsoft/clarity-tools.git"
+    "url": "git+https://github.com/Microsoft/clarity-tools.git"
   },
   "bugs": {
     "url": "https://github.com/Microsoft/clarity-tools/issues"
   },
   "dependencies": {
+    "clarity": "^1.0.2",
```
Comment (ender336) [inline package.json]: > "clarity": "^1.0.2", [](start = 3, length = 21)

why was this added? 

[https://www.npmjs.com/package/clarity](https://www.npmjs.com/package/clarity)

That's not our package

File: package.json [hunk 2]
```diff
@@ -51,10 +53,11 @@
...
   },
   "scripts": {
-    "compile": "node_modules/.bin/tsc -p ./",
-    "bundle:content": "node_modules/.bin/browserify ./extension/instrument.js > ./extension/content.js",
-    "bundle:visualization": "node_modules/.bin/browserify ./extension/app.js > ./extension/clarity.js",
-    "binplace:extension": "node_modules/.bin/copyfiles -f ./src/extension/*.html ./src/extension/manifest.json ./src/extension/*.png ./extension/",
+    "compile": "tsc -p ./",
+    "bundle:content": "browserify ./extension/instrument.js > ./extension/content.js",
+    "bundle:visualization": "browserify ./extension/app.js > ./extension/clarity.js",
+    "binplace:extension": "copyfiles -f ./src/extension/*.html ./src/extension/manifest.json ./src/extension/*.png ./extension/",
```
Comment (ender336) [inline package.json]: why are these changing? My path doesn't include node_modules/,bin so these commands would fail now

File: package.json [hunk 3]
```diff
@@ -18,18 +18,20 @@
...
   "dependencies": {
+    "clarity": "^1.0.2",
     "clarity-js": "^0.1.29",
     "compare-versions": "^3.1.0",
     "material-ui": "0.17.4",
+    "query-string": "^6.0.0",
     "react": "^15.6.2",
     "react-dom": "^15.6.2",
-    "react-redux": "^5.0.6",
+    "react-redux": "^5.0.7",
```
Comment (ender336) [inline package.json]: >"^ [](start = 19, length = 2)

doesn't the caret already handle minor versions?

File: src/extension/app.tsx [hunk 4]
```diff
@@ -51,6 +51,7 @@ chrome.runtime.sendMessage({ fetch: true }, function (response) {
             if (!(id in structuredEvents)) {
                 structuredEvents[id] = { envelope: json.envelope, events: [] };
                 structuredEvents[id]["envelope"].dateTime = entry.dateTime;
+                //structuredEvents[id]["envelope"].PLTV2 = 123; // LOGIC ;
```
Comment (ender336) [inline src/extension/app.tsx]: > //structuredEvents[id]["envelope"].PLTV2 = 123; // LOGIC ; [](start = 15, length = 59)

revert this file

File: src/visualization/actions/index.ts [hunk 5]
```diff
@@ -9,7 +9,9 @@ export const enum Types {
     ToggleFullPage,
     NotFound,
     ToggleInactiveSession,
-    Error
+    Error,
+    ShowSettleEvents,
```
Comment (ender336) [inline src/visualization/actions/index.ts]: >ShowSettleEvents [](start = 4, length = 16)

as discussed offline, let's move away from calling this settle events or settle time. It would be better branded as perceivedLoadTime. 

For this particular Types definition, maybe ShowPltVizualization ?

File: src/visualization/components/Content.tsx [hunk 6]
```diff
@@ -3,13 +3,16 @@ import Replay from "./Replay";
 import { connect } from "react-redux";
 import { Tabs, Tab } from 'material-ui/Tabs';
 import SwipeableViews from 'react-swipeable-views';
-import { selectView } from "../actions";
+import { selectView , showSettleEvents} from "../actions";
 import { bindActionCreators } from "redux";
 import Header from "./Header";
+import Snapshot from "./Snapshot";
```
Comment (ender336) [inline src/visualization/components/Content.tsx]: >Snapshot [](start = 7, length = 8)

we don't need this right?

File: src/visualization/components/Content.tsx [hunk 7]
```diff
@@ -19,6 +22,7 @@ class Content extends React.Component<any, any> {
         let Content = this.props.notfound ? <div className={'clarity-notfound'}>{notfound}</div> : <Replay />;
 
         if (this.props.impression != null) {
+            var newImpression = { envelope: {...this.props.impression.envelope}, events: this.props.impression.events.slice()};
```
Comment (ender336) [inline src/visualization/components/Content.tsx]: >newImpression [](start = 16, length = 13)

not used

File: src/visualization/components/Content.tsx [hunk 8]
```diff
@@ -53,5 +59,9 @@ class Content extends React.Component<any, any> {
 // mapStateToProps and matchDispatchToProps using fat arrow function
 export default connect(
     state => { return { impression: state.impression, notfound: state.notfound, menu: state.menu, view: state.view } },
-    dispatch => { return bindActionCreators({ selectView: selectView }, dispatch) }
+    dispatch => { return bindActionCreators({ 
+        selectView: selectView, 
+        showSettleEvents: showSettleEvents
```
Comment (ender336) [inline src/visualization/components/Content.tsx]: >showSettleEvents [](start = 26, length = 16)

this doesn't actually do anything does it? onChange is still calling selectView

File: src/visualization/components/Player.tsx [hunk 9]
```diff
@@ -17,8 +17,11 @@ class Player extends React.Component<any, any> {
     private setTimeoutId = -1;
     private activeImpressionId = "";
     private keyFrames = [];
+    private redEvents = [];
+
```
Comment (ender336) [inline src/visualization/components/Player.tsx]: extra whitespace/newline

File: src/visualization/components/Player.tsx [hunk 10]
```diff
@@ -66,7 +69,9 @@ class Player extends React.Component<any, any> {
     }
 
     showFullPage() {
+        //console.log(this.props.fullpage);
         this.props.showFullPage(!this.props.fullpage);
```
Comment (ender336) [inline src/visualization/components/Player.tsx]: remove

File: src/visualization/components/Player.tsx [hunk 11]
```diff
@@ -101,9 +108,11 @@ class Player extends React.Component<any, any> {
         let iconTooltip = this.props.playback ? "Pause playback" : "Start playback";
         let speedIconColor = this.props.speed ? "white" : "#666";
         let fullPageIconColor = this.props.fullpage ? "white" : "#666";
+        
         let prevIconColor = index > 0 ? "white" : "#666";
         let nextIconColor = index < (this.props.playlist.length - 1) ? "white" : "#666";
         
+
```
Comment (ender336) [inline src/visualization/components/Player.tsx]: remove

File: src/visualization/components/Player.tsx [hunk 12]
```diff
@@ -101,9 +108,11 @@ class Player extends React.Component<any, any> {
         let iconTooltip = this.props.playback ? "Pause playback" : "Start playback";
         let speedIconColor = this.props.speed ? "white" : "#666";
         let fullPageIconColor = this.props.fullpage ? "white" : "#666";
+
```
Comment (ender336) [inline src/visualization/components/Player.tsx]: remove

File: src/visualization/components/Player.tsx [hunk 13]
```diff
@@ -156,14 +165,16 @@ export default connect(
             playback: state.playback,
             speed: state.speed,
             view: state.view,
-            fullpage: state.fullpage            
+            fullpage: state.fullpage,
+
```
Comment (ender336) [inline src/visualization/components/Player.tsx]: unnecessary, same below

File: src/visualization/components/Player.tsx [hunk 14]
```diff
@@ -17,8 +17,11 @@ class Player extends React.Component<any, any> {
     private setTimeoutId = -1;
     private activeImpressionId = "";
     private keyFrames = [];
+    private redEvents = [];
```
Comment (ender336) [inline src/visualization/components/Player.tsx]: > private redEvents = []; [](start = 3, length = 24)

never used?

File: src/visualization/components/Snapshot.tsx [hunk 15]
```diff
@@ -5,6 +5,9 @@ import Layout from "../parsers/layout";
 import BoxModel from "../parsers/boxmodel";
 import Viewport from "../parsers/viewport";
 import Pointer from "../parsers/pointer";
+import SettleLayout from  "../parsers/settleLayout";
+import {showSettleEvents} from "../actions";
```
Comment (ender336) [inline src/visualization/components/Snapshot.tsx]: never used?

File: src/visualization/components/Snapshot.tsx [hunk 16]
```diff
@@ -28,9 +32,14 @@ class Snapshot extends React.Component<any, any> {
       var events = this.props.impression.events;
       var start = events[0].time;
       var end = events[events.length - 1].time;
+      
+
```
Comment (ender336) [inline src/visualization/components/Snapshot.tsx]: whitespace

File: src/visualization/components/Snapshot.tsx [hunk 17]
```diff
@@ -46,19 +55,45 @@ class Snapshot extends React.Component<any, any> {
...
+            //console.log("Parser", parser);
+ 
           }
+          
+
           this.currentPointer = i;
         }
         else break;
       }
+      //console.log("here inside snapshot", this.redEvents);
```
Comment (ender336) [inline src/visualization/components/Snapshot.tsx]: >//console.log("here inside snapshot", this.redEvents); [](start = 6, length = 54)

remove comments and your debugging console.logs

File: src/visualization/components/Snapshot.tsx [hunk 18]
```diff
@@ -20,6 +23,7 @@ class Snapshot extends React.Component<any, any> {
   private activeView = 0;
   private currentTime = 0;
   private currentPointer = -1;
+  private redEvents = this.props.impression;
```
Comment (ender336) [inline src/visualization/components/Snapshot.tsx]: >redEvents [](start = 10, length = 9)

not used other than your debug stuff

File: src/visualization/components/Snapshot.tsx [hunk 19]
```diff
@@ -46,19 +55,45 @@ class Snapshot extends React.Component<any, any> {
...
+          //let parser = event.type === "Layout" && this.props.view == 1 ? "BoxModel" : event.type;
+          let parser = event.type;
+          if(event.type === "Layout"){
+            if(this.props.view == 1){
+              parser = "BoxModel";
+            }else if(this.props.view == 2){
+              parser = "SettleEvents";
+            }else{
+              parser = event.type;
+            }
```
Comment (ender336) [inline src/visualization/components/Snapshot.tsx]: your else is misplaced here - you want to default to event.type in cases where type was not layout as well. It would be cleanest to not nest your if statements

File: src/visualization/parsers/ISettleElementLayoutState.ts [hunk 20]
```diff
@@ -0,0 +1,5 @@
+import { IAttributes, ILayoutState, IElementLayoutState, IDoctypeLayoutState, ITextLayoutState, IIgnoreLayoutState, Action } from "clarity-js/clarity";
```
Comment (ender336) [inline src/visualization/parsers/ISettleElementLayoutState.ts]: >IElementLayoutState [](start = 36, length = 19)

you're only using this

File: src/visualization/parsers/layout.ts [hunk 21]
```diff
@@ -193,6 +195,8 @@ export default class Layout implements IParser {
                     img.style.width = state.layout.width + "px";
                     img.style.height = state.layout.height + "px";
                 }
+                
+
```
Comment (ender336) [inline src/visualization/parsers/layout.ts]: revert this file

File: src/visualization/parsers/settleLayout.ts [hunk 22]
```diff
@@ -0,0 +1,267 @@
+/// <reference path="../../../node_modules/clarity-js/clarity.d.ts" />
+import { IParser } from "../components/Snapshot";
+import { IAttributes, ILayoutState, IElementLayoutState, IDoctypeLayoutState, ITextLayoutState, IIgnoreLayoutState, Action } from "clarity-js/clarity";
+import {ISettleElementLayoutState} from "./ISettleElementLayoutState";
+export default class Layout implements IParser {
```
Comment (ender336) [inline src/visualization/parsers/settleLayout.ts]: >Layout [](start = 21, length = 6)

99% of this is just a copy of the other layout right? Let's extend the class and just override whatever small bits are changing

File: src/visualization/reducers/impression.ts [hunk 23]
```diff
@@ -10,16 +10,66 @@ export default function (state = null, action) {
...
+            return getSettleEvents(action.payload);
     }
     return state;
 }
 
+function getSettleEvents(impression) {
+
+    return settleTime(sort(impression));
+
+}
```
Comment (ender336) [inline src/visualization/reducers/impression.ts]: let's remove this level of indirection, it doesn't add anything

File: src/visualization/reducers/index.ts [hunk 24]
```diff
@@ -24,7 +25,8 @@ const reducers = combineReducers({
     inactive: InactiveReducer,
     error: ErrorReducer,
     fullpage: FullPageReducer,
-    notfound: NotFoundReducer
+    notfound: NotFoundReducer,
+    
 });
```
Comment (ender336) [inline src/visualization/reducers/index.ts]: revert this file

File: src/visualization/components/Content.tsx (hunk 8 above)
Comment (lin99) [reply src/visualization/components/Content.tsx]: The select view was there before, but the ShowSEttleEvents is indeed not useful here. 

---
In reply to: [175850712](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175850712) [](ancestors = 175850712)

File: src/extension/app.tsx (hunk 4 above)
Comment (ender336) [reply src/extension/app.tsx]: not fixed yet

---
In reply to: [175848321](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175848321) [](ancestors = 175848321)

File: package.json [hunk 25]
```diff
@@ -27,6 +27,7 @@
     "clarity-js": "^0.1.29",
     "compare-versions": "^3.1.0",
     "material-ui": "0.17.4",
+    "query-string": "^6.0.0",
```
Comment (ender336) [inline package.json]: >  "query-string": "^6.0.0", [](start = 2, length = 27)

let's remove this and our reference to it in app.tsx - we don't actually use that parse anywhere

File: src/visualization/actions/index.ts (hunk 5 above)
Comment (ender336) [reply src/visualization/actions/index.ts]: not resolved - if you disagree mark as won'tfix and give an explanation

---
In reply to: [175850109](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175850109) [](ancestors = 175850109)

File: src/visualization/components/Player.tsx (hunk 12 above)
Comment (ender336) [reply src/visualization/components/Player.tsx]: not resolved - if you disagree mark as won'tfix and give an explanation

---
In reply to: [175851011](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175851011) [](ancestors = 175851011)

File: src/visualization/components/Player.tsx (hunk 11 above)
Comment (ender336) [reply src/visualization/components/Player.tsx]: not resolved - if you disagree mark as won'tfix and give an explanation

---
In reply to: [175850963](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175850963) [](ancestors = 175850963)

File: src/visualization/components/Player.tsx (hunk 13 above)
Comment (ender336) [reply src/visualization/components/Player.tsx]: not resolved - if you disagree mark as won'tfix and give an explanation

---
In reply to: [175851128](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175851128) [](ancestors = 175851128)

File: src/visualization/components/Snapshot.tsx [hunk 26]
```diff
@@ -1,10 +1,13 @@
...
 import * as ReactDOM from "react-dom";
 import { connect } from "react-redux";
-import Layout from "../parsers/layout";
+import { Layout } from "../parsers/layout";
 import BoxModel from "../parsers/boxmodel";
 import Viewport from "../parsers/viewport";
 import Pointer from "../parsers/pointer";
+import PerceivedLoadTimeLayout from  "../parsers/perceivedLoadTimeLayout";
+
+
```
Comment (ender336) [inline src/visualization/components/Snapshot.tsx]: get rid of the extra whitespace you added

File: src/visualization/components/Snapshot.tsx (hunk 16 above)
Comment (ender336) [reply src/visualization/components/Snapshot.tsx]: not resolved - if you disagree mark as won'tfix and give an explanation

---
In reply to: [175851523](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175851523) [](ancestors = 175851523)

File: src/visualization/parsers/layout.ts [hunk 27]
```diff
@@ -133,10 +133,12 @@ export default class Layout implements IParser {
                 this.attributes(img, state.attributes);
                 if (!img.src)
                 {
+                    
                     img.src = this.placeholderImage;
                     img.style.width = state.layout.width + "px";
                     img.style.height = state.layout.height + "px";
                 }
+
```
Comment (ender336) [inline src/visualization/parsers/layout.ts]: remove these whitespace lines

File: src/visualization/parsers/layout.ts (hunk 21 above)
Comment (ender336) [reply src/visualization/parsers/layout.ts]: not resolved - if you disagree mark as won'tfix and give an explanation

---
In reply to: [175853297](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175853297) [](ancestors = 175853297)

File: src/visualization/parsers/perceivedLoadTimeLayout.ts [hunk 28]
```diff
@@ -0,0 +1,185 @@
...
+                        let textNode = document.createTextNode(state.cssRules[i]);
+                        node.appendChild(textNode);
+                    }
+                }
+                this.layouts[state.index] = this.domInsert(node, parent, next);
+                break;
+        }
+    }
+
+    private updatePerceivedLayout(state: IPerceivedElementLayoutState) {
```
Comment (ender336) [inline src/visualization/parsers/perceivedLoadTimeLayout.ts]: >private [](start = 4, length = 7)

let's keep these methods as protected to be consistent with the parent class

File: src/visualization/parsers/perceivedLoadTimeLayout.ts [hunk 29]
```diff
@@ -0,0 +1,185 @@
...
+        this.document = document;
+        this.base = base;
+    }
+
+    private insertPerceivedLayout (layoutState: ILayoutState) {
+        var doc = this.document;
+        var state: any = layoutState as IPerceivedElementLayoutState;
+        var parent = this.layouts[state.parent];
+        var next = state.next in this.layouts ? this.layouts[state.next] : null;
+        switch (state.tag) {
```
Comment (ender336) [inline src/visualization/parsers/perceivedLoadTimeLayout.ts]: >switch (state.tag) { [](start = 8, length = 20)

huge parts of this file are still copy pasted from layout.ts. Can we make a protected helper method which is a noop in layout.ts but does your small bit of extra work for the child class?

File: src/visualization/reducers/impression.ts [hunk 30]
```diff
@@ -10,16 +10,61 @@ export default function (state = null, action) {
...
+                    break;
+                }
+                lastTime = evt.time;
+            }else{
+                continue;
+            }
+ 
+        }
+
+        console.log("LAST TIME", lastTime);
```
Comment (ender336) [inline src/visualization/reducers/impression.ts]: > console.log("LAST TIME", lastTime); [](start = 7, length = 36)

remove

File: src/visualization/reducers/index.ts (hunk 24 above)
Comment (ender336) [reply src/visualization/reducers/index.ts]: not resolved - if you disagree mark as won'tfix and give an explanation

---
In reply to: [175853932](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r175853932) [](ancestors = 175853932)

File: src/visualization/actions/index.ts [hunk 31]
```diff
@@ -9,7 +9,8 @@ export const enum Types {
     ToggleFullPage,
     NotFound,
     ToggleInactiveSession,
-    Error
+    Error,
+
```
Comment (ender336) [inline src/visualization/actions/index.ts]: we can revert this file now

File: src/visualization/components/Content.tsx [hunk 32]
```diff
@@ -27,10 +29,12 @@ class Content extends React.Component<any, any> {
                         <Tabs className="clarity-tabs" onChange={this.props.selectView} value={this.props.view} tabItemContainerStyle={{backgroundColor: "#666"}}>
                             <Tab label="Replay" value={0} />
                             <Tab label="Box Model" value={1} />
+                            <Tab label="Settle Images" value={2} />
```
Comment (ender336) [inline src/visualization/components/Content.tsx]: >Settle Images [](start = 40, length = 13)

Can we rename this? Maybe something simple like "Page Performance"?

File: src/visualization/components/Snapshot.tsx [hunk 33]
```diff
@@ -46,20 +48,28 @@ class Snapshot extends React.Component<any, any> {
...
       }
-      
+
       var startPointer = time < this.currentTime ? 0 : this.currentPointer + 1;
       for (var i = startPointer; i < events.length; i++) {
         var event = events[i];
         if (event.time <= time) {
-          let parser = event.type === "Layout" && this.props.view == 1 ? "BoxModel" : event.type; 
+          
+          //let parser = event.type === "Layout" && this.props.view == 1 ? "BoxModel" : event.type;
```
Comment (ender336) [inline src/visualization/components/Snapshot.tsx]: >//let parser = event.type === "Layout" && this.props.view == 1 ? "BoxModel" : event.type; [](start = 10, length = 89)

remove commented code please

File: src/visualization/parsers/perceivedLoadTimeLayout.ts [hunk 34]
```diff
@@ -0,0 +1,45 @@
...
+    
+    setup(document: Document, frame: HTMLIFrameElement, base: string, thumbnail? : boolean) {
+        this.layouts = {};
+        this.document = document;
+        this.base = base;
+    }
+
+    protected insertHelper(state: IPerceivedElementLayoutState, img : HTMLImageElement ){
+        if(state.isPerceivedEvent === false ){
+            img.style.color = "red";
```
Comment (ender336) [inline src/visualization/parsers/perceivedLoadTimeLayout.ts]: >img.style.color = "red"; [](start = 12, length = 24)

I've noticed the opacity, but haven't noticed the color change when I've seen the visualizations. Is changing the color doing anything? If so this is fine, but if not we should just remove it

File: src/visualization/reducers/impression.ts [hunk 35]
```diff
@@ -22,4 +22,38 @@ function sort(impression) {
...
+        }
+
+        for (var evt of impression.events) {
+            if(evt.type === "Layout" && evt.state.tag === "IMG" && 
+                evt.state.layout.width >= mindWidth && evt.state.layout.height >= minHeight && evt.time <= lastTime ){
+                   
+                    evt["state"]["isPerceivedEvent"] = true;
+            } else{
+                evt["state"]["isPerceivedEvent"] = false;
+            }
```
Comment (ender336) [inline src/visualization/reducers/impression.ts]: do we need to explicitly set this to false? If it isn't on the object at all we can trust it wasn't a perceivedperf event right?

Comment (lin99) [reply src/visualization/reducers/impression.ts]: We can do that but I set it up to false to avoid undefined exceptions at any moment. Also the condition on change the opacity of the event is under the false value.  

---
In reply to: [177475290](https://github.com/Microsoft/clarity-tools/pull/41#discussion_r177475290) [](ancestors = 177475290)
